src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -249,3 +438,221 @@ def calculate_ramp_position_and_duration(
+    # We do this by substracting against previous cumulative time
+    time_intervals = np.diff(np.concatenate(([0.0], combined_times))).tolist()
+
+    # Prepare dicts for the resulting position and velocity profiles over the turnaround
+    turnaround_profile: dict[Motor, npt.NDArray[np.float64]] = {}
+    turnaround_velocity: dict[Motor, npt.NDArray[np.float64]] = {}
+
+    # Loop over each motor and integrate its velocity profile over the unified times
+    for motor in motors:
+        axis_times = time_arrays[motor]
+        axis_velocities = velocity_arrays[motor]
+        axis_position = slice.upper[motor][
+            gap - 1
+        ]  # start position at beginning of the gap
+        prev_vel = axis_velocities[0]  # last velocity seen (start at initial velocity)
+        time_since_prev_axis_point = 0.0  # elapsed time since the last velocity point
+        axis_idx = 1  # index into this axis's velocity/time arrays
+
+        # Allocate output arrays for this motor with correct size
+        turnaround_profile[motor] = np.empty(num_intervals, dtype=np.float64)
+        turnaround_velocity[motor] = np.empty(num_intervals, dtype=np.float64)
+
+        # Step through each interval in the Δt list
+        for i, dt in enumerate(time_intervals):
+            next_vel = axis_velocities[axis_idx]
+            prev_axis_vel = axis_velocities[axis_idx - 1]
+            axis_dt = axis_times[axis_idx] - axis_times[axis_idx - 1]
+
+            if np.isclose(combined_times[i], axis_times[axis_idx]):
+                # If the current combined time exactly matches this motor's
+                # next velocity change point, no interpolation is needed, so:
+                this_vel = next_vel
+                axis_idx += 1
+                time_since_prev_axis_point = 0.0
+            else:
+                # Otherwise, linearly interpolate velocity between the previous
+                # and next known velocity points for this motor
+                time_since_prev_axis_point += dt
+                # The fraction of the way we are from previous to next known velocities
+                # for this motor
+                frac = time_since_prev_axis_point / axis_dt
+                # Interpolate for our velocity
+                this_vel = prev_axis_vel + frac * (next_vel - prev_axis_vel)
+
+            # Integrate velocity over this interval to update position.
+            # Using the trapezoidal rule:
+            delta_pos = 0.5 * (prev_vel + this_vel) * dt
+            axis_position += delta_pos
+            prev_vel = this_vel  # update for next loop
+
+            # Store the computed position and velocity for this interval
+            turnaround_profile[motor][i] = axis_position
+            turnaround_velocity[motor][i] = this_vel
+
+    return GapSegment(
+        turnaround_profile,
+        turnaround_velocity,
+        time_intervals,
+        len(turnaround_profile[motors[0]]),  # Motors have same gap lengths
```

Comment:
Could you pass by keyword here so I don't have to scroll up to see what each parameter is?